hook.c

```diff
@@ -73,6 +74,58 @@ static struct HookList Hooks = TAILQ_HEAD_INITIALIZER(Hooks);
 static struct HashTable *IdxFmtHooks = NULL;
 static HookFlags current_hook_type = MUTT_HOOK_NO_FLAGS;
 
+/**
+ * mutt_parse_charset_iconv_hook - Parse 'charset-hook' and 'iconv-hook' commands - Implements Command::parse() - @ingroup command_parse
+ */
+enum CommandResult mutt_parse_charset_iconv_hook(struct Buffer *buf, struct Buffer *s,
+                                                 intptr_t data, struct Buffer *err)
+{
+  struct Buffer *alias = mutt_buffer_pool_get();
+  struct Buffer *charset = mutt_buffer_pool_get();
+
+  int rc = MUTT_CMD_ERROR;
+  int retval = 0;
+
+  enum LookupType type = MUTT_LOOKUP_CHARSET;
+  if (data & MUTT_ICONV_HOOK)
+  {
+    type = MUTT_LOOKUP_ICONV;
+  }
+
+  retval += mutt_extract_token(alias, s, MUTT_TOKEN_NO_FLAGS);
+  retval += mutt_extract_token(charset, s, MUTT_TOKEN_NO_FLAGS);
+
+  if (retval != 0)
+  {
+    mutt_buffer_pool_release(&alias);
+    mutt_buffer_pool_release(&charset);
+    return MUTT_CMD_ERROR;
+  }
+
+  if (mutt_buffer_is_empty(alias) || mutt_buffer_is_empty(charset))
+  {
+    mutt_buffer_printf(err, _("%s: too few arguments"), buf->data);
+    rc = MUTT_CMD_WARNING;
+  }
+  else if (mutt_ch_lookup_add(type, mutt_buffer_string(alias),
+                              mutt_buffer_string(charset), err))
+  {
+    rc = MUTT_CMD_SUCCESS;
+  }
+
+  if (MoreArgs(s))
+  {
+    mutt_buffer_printf(err, _("%s: too many arguments"), buf->data);
+    mutt_buffer_reset(s); // clean up buffer to avoid a mess with further rcfile processing
+    rc = MUTT_CMD_WARNING;
+  }
+
```

Comment:
We could put a `done:` label here

Reply:
I generally avoid Labels and goto.
```Go To Statement Considered Harmful```

It would be better to allocate the temp buffers from the caller, and give them as parameters to the parsing function.

I think, mutt_parse_hook() could be reestablished as the entry point into the parsing, looks up the data-flags, allocates the temp buffers, calling the sub-parsers with these args added and deallocates the temp buffers.
This would make things cleaner: the sub-paresrs don't need to do the alloaction and deallocation.

Reply:
Yes, I'm familiar with Dijkstra :-) however...
In C, it's common to reduce complexity by grouping all the cleanup in one place -- at the end.

When we move to C++, life will be simpler
(Don't ask, it's not going to be any time soon).

Reply:
> Yes, I'm familiar with Dijkstra :-) however... In C, it's common to reduce complexity by grouping all the cleanup in one place -- at the end.

1) goto not needed for cleanup at the end
2) if such constructs like the goto's are needed, IMO that's a smell.
3) is it "common" in C? Not sure, this is  a language issue.

See my new commit, how the return value is gathered from the call to the subparser (and the allocation/cleanup is completely out of the sup-barsers - this one and potentially all others that will be implemented in the future).
Doing things similar for the other sub-parsers would make the code much cleaner.
if/else instead of goto.
Even more abstraction would be possible (function pointers...).

> 
> When we move to C++, life will be simpler (Don't ask, it's not going to be any time soon).

I'm completely happy with C. And using, what I proposed could offer enough potential for abstraction, even just with C. No goto needed. And no C++ needed :-)